internal/mode/static/state/graph/backend_refs.go

```diff
@@ -288,16 +307,39 @@ func getServiceAndPortFromRef(
 
 	svc, ok := services[svcNsName]
 	if !ok {
-		return svcNsName, v1.ServicePort{}, field.NotFound(refPath.Child("name"), ref.Name)
+		return svcNsName, v1.ServicePort{}, []v1.IPFamily{}, field.NotFound(refPath.Child("name"), ref.Name)
 	}
 
 	// safe to dereference port here because we already validated that the port is not nil in validateBackendRef.
 	svcPort, err := getServicePort(svc, int32(*ref.Port))
 	if err != nil {
-		return svcNsName, v1.ServicePort{}, err
+		return svcNsName, v1.ServicePort{}, []v1.IPFamily{}, err
 	}
 
-	return svcNsName, svcPort, nil
+	svcIPFamilies := svc.Spec.IPFamilies
+
+	return svcNsName, svcPort, svcIPFamilies, nil
+}
+
+func verifyIPFamily(npCfg *NginxProxy, svcIPFamily []v1.IPFamily) error {
+	if npCfg.Source == nil {
+		return nil
+	}
+
+	// we can access this field since we have already validated that ipFamily is not nil in validateNginxProxy.
+	npIPFamily := npCfg.Source.Spec.IPFamily
+	if *npIPFamily == ngfAPI.IPv4 {
+		if slices.Contains(svcIPFamily, v1.IPv6Protocol) {
+			return fmt.Errorf("service configured with IPv6 stack but NGINX only supports IPv4")
```

Comment:
suggest using NginxProxy instead of NGINX here, so it is more clear where the mismatch can be resolved. same for the error below
```suggestion
			return fmt.Errorf("service configured with IPv6 family but NginxProxy is configured with IPv4")
```